pennylane/math/ode.py

```diff
@@ -0,0 +1,246 @@
+        host_callback.id_tap(
+            _tolerance_warn,
+            (
+                atol,
+                rtol,
+                mean_err_ratio,
+                jnp.sqrt(jnp.mean(_abs2(y1_error))),
+                jnp.sqrt(jnp.mean(_abs2(err_tol))),
+            ),
+        )
+
+        carry = [y1, f1, t1]
+        return carry, y1
+
+    # scan = fancy for-loop in jax, see https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html
+    f0 = func_(y0, ts[0])
+    init = [y0, f0, ts[0]]
+    carry, _ = jax.lax.scan(scan_fun, init, ts[1:])
+    y, _, _ = carry
+    # we typically only care about the final U
+    # In case we want to return all ys, the output of
+    # odeint() must be vmapped, jax.vmap(unravel)(out)
+    return y  # jnp.concatenate((y0[None], ys))
+
+
+def runge_kutta_step(func, y0, f0, t0, dt):
+    """Performing the Runge Kutta 45 step
+
+    Butcher table taken from `wikipedia.org/Dormand Prince <https://en.wikipedia.org/wiki/Dormand%E2%80%93Prince_method>`_
+    """
+    alpha = jnp.array([1 / 5, 3 / 10, 4 / 5, 8 / 9, 1.0, 1.0, 0], dtype=dt.dtype)
+    beta = jnp.array(
+        [
+            [1 / 5, 0, 0, 0, 0, 0, 0],
+            [3 / 40, 9 / 40, 0, 0, 0, 0, 0],
+            [44 / 45, -56 / 15, 32 / 9, 0, 0, 0, 0],
+            [19372 / 6561, -25360 / 2187, 64448 / 6561, -212 / 729, 0, 0, 0],
+            [9017 / 3168, -355 / 33, 46732 / 5247, 49 / 176, -5103 / 18656, 0, 0],
+            [35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84, 0],
+        ],
+        dtype=f0.dtype,
+    )
+    c_sol = jnp.array(
+        [35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84, 0], dtype=f0.dtype
+    )
+
+    # TODO: include error check
+    c_error = jnp.array(
+        [
+            35 / 384 - 1951 / 21600,
+            0,
+            500 / 1113 - 22642 / 50085,
+            125 / 192 - 451 / 720,
+            -2187 / 6784 - -12231 / 42400,
+            11 / 84 - 649 / 6300,
+            -1.0 / 60.0,
+        ],
+        dtype=f0.dtype,
+    )
+
+    def body_fun(i, k):
+        ti = t0 + dt * alpha[i - 1]
+        yi = y0 + dt.astype(f0.dtype) * jnp.dot(beta[i - 1, :], k)
+        ft = func(yi, ti)
+        return k.at[i, :].set(ft)
+
+    k = jnp.zeros((7, f0.shape[0]), f0.dtype).at[0, :].set(f0)
+    k = jax.lax.fori_loop(1, 7, body_fun, k)
+
+    y1 = dt.astype(f0.dtype) * jnp.dot(c_sol, k) + y0
+    y1_error = dt.astype(f0.dtype) * jnp.dot(c_error, k)  # see coeffs, this is y(coeff) - y(coeff*)
+    f1 = k[-1]
+    return y1, f1, y1_error
+
+
+# pylint: disable=no-member
+def ravel_first_arg(f, unravel):
+    """ "Decorate a function to work with a raveled first argument"""
+    return _ravel_first_arg(lu.wrap_init(f), unravel).call_wrapped
+
+
+@lu.transformation
+def _ravel_first_arg(unravel, y_flat, *args):
+    y = unravel(y_flat)
+    ans = yield (y,) + args, {}
+    ans_flat, _ = ravel_pytree(ans)
+    yield ans_flat
```

Comment:
When returning function wrappers, we usually put the wrapper function inside the outer function. This way it is clear that the wrapper function (`_ravel_first_arg`) is only used inside the outer function (`ravel_first_arg`)